controllers/collection.lua

```diff
@@ -87,53 +78,106 @@ CollectionController.GET.user_collections = function (self)
     -- Description: Get info about a collection.
     -- Parameters:  username, collection_slug, ...
 
-    return -- TODO
-    -- local collection = assert_collection_exists(self)
-    -- local project_count = collection:count_projects
-    -- collection.projects_count = project_count
-    -- return jsonResponse(collection)
+    local collection = assert_collection_exists(self)
+    collection.projects_count = collection:count_projects()
+    return jsonResponse(collection)
 end
 
-CollectionController.POST.collection = function (self)
+CollectionController.POST.collections = json_params(function (self)
     -- POST /users/:username/collections/:collection_slug
     -- Description: Create a collection.
     -- Parameters:  username, collection_name, ...
 
-    -- TODO (temp off): assert_all({ assert_logged_in, assert_users_match }, self)
-    -- Must assert name before generating a slug.
-    validate.assert_valid(self.params, { { 'name', exists = true } })
+    assert_users_match(self)
+    local params = self.params
+    local collection = Collections:find(self.queried_user.id,
+                                        params.collection_slug)
+
+    if collection then
+        -- TODO: I think we can extract these into functions.
+        local published = params.published ~= nil and params.published == true
+        local published_at = (published and collection.published_at) or
+            current_time_or_nil(published)
+        local shared = params.shared ~= nil and params.shared == true
+        local shared_at = (shared and collection.shared_at) or
+             current_time_or_nil(shared)
+
+        collection:update({
+            name = params.name or collection.name,
+            slug = params.name and util.slugify(params.name) or collection.slug,
+            description = params.description or collection.description,
+            published = published,
+            published_at = published_at,
+            shared = shared,
+            shared_at = shared_at,
+            thumbnail_id = params.thumbnail_id or collection.thumbnail_id
+        })
+
+        return jsonResponse(collection)
+    end
 
+    -- Must assert name before generating a slug.
+    validate.assert_valid(params, { { 'name', exists = true } })
     return jsonResponse(assert_error(Collections:create({
-        name = self.params.name,
-        slug = util.slugify(self.params.name),
-        creator_id = self.request_user.id,
-        description = self.params.description,
-        published = self.params.published,
-        published_at = current_time_or_nil(self.params.published),
-        shared = self.params.shared,
-        shared_at = current_time_or_nil(self.params.shared),
-        thumbnail_id = self.params.thumbnail_id
+        name = params.name,
+        slug = util.slugify(params.name),
+        creator_id = queried_user.id,
+        description = params.description,
+        published = params.published  == true,
```

Comment:
Are you sure booleans come as booleans in params? I think I recall they come in as strings. We could always add a before filter turning `'true'` and `'false'` into actual bools, but then we'll need to hunt all the places where we're checking against these strings. Not a big hassle, just something to remember.